Upgrade PyPI development dependencies

U8NWXD opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #[n/a].
2. This PR does the following: Upgrade development dependencies from PyPI

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

No proof of changes needed because these development dependencies only affect the developer workflow, so they should not affect how the Oppia server runs. The opening of this PR and CI checks show that the developer workflow is working with these changes.

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

U8NWXD committed:
Upgrade PyPI Development Dependencies

Changelogs for the upgraded dependencies in `requirements_dev.in` were
skimmed. After this commit is merged, developers should run `python -m
scripts.install_python_dev_dependencies` to make sure their
installations are up-to-date.

U8NWXD committed:
Upgrade a Few PyPI Prod Dependencies

Upgrade only the PyPI prod dependencies that were also upgraded as part
of the dev dependencies upgrade. After this commit is merged, developers
should run `python -m scripts.install_third_party` to make sure they
have the updated dependencies.

U8NWXD committed:
Remove obsolete TODO

U8NWXD committed:
Update pylint_extensions.py for latest pylint

U8NWXD committed:
Update pylint config for new pylint version

U8NWXD committed:
Fix docstring validity checks and linting

U8NWXD committed:
Allow assigning lambdas to variables

U8NWXD committed:
Fix docstrings_checker.py tests

seanlip commented on .pylintrc:
```diff
@@ -158,6 +152,8 @@ disable=abstract-method,
         # broken by putting imports inside functions so they aren't
         # executed upon module import.
         cyclic-import,
+        broad-exception-raised,
```
Why are we adding new exceptions?

U8NWXD replied:
This rule disallows raising exceptions of type `Exception`, which we do frequently in the code base. This must be a new lint rule introduced since we last upgraded pylint. I don't think it's worth changing all the exception types, so I'm ignoring the rule instead

seanlip reviewed (changes requested):
Thanks @U8NWXD! Just one comment.

Also CI checks are failing, FYI.

seanlip reviewed (approved):
OK, thanks! LGTM.

gp201 reviewed (changes requested):
Thanks, @U8NWXD !
Just one question. What is the plan for the redis upgrade?

U8NWXD commented:
@gp201 I haven't gotten to redis yet. I'm starting with the dependencies I'm most familiar with since this is my first time upgrading Oppia's dependencies

gp201 reviewed (approved):
LGTM

gp201 commented:
@U8NWXD FYI, the lint checks seem to be failing.

U8NWXD committed:
Fix lint errors

U8NWXD committed:
Upgrade pip

lkbhitesh07 reviewed (approved):
The changes look good to me. Just FYI the backend tests are failing.
Thanks

U8NWXD committed:
Reinstall installation tools during build

aasiffaizal reviewed (approved):
lgtm

U8NWXD committed:
Add core.tests.test_utils_test to coverage ignore

Before the dependency upgrade, coverage reports from backend test runs
on develop show that core.tests.test_utils_test was not being executed
for some reason. With the upgrade, it is now being included in the test
suite, but since it does not have complete per-file coverage, it needs
to be added to the ignore list.

U8NWXD committed:
Fix backend tests

U8NWXD committed:
Merge branch 'develop' into upgrade-pypi-dev

U8NWXD committed:
Fix protoc version

U8NWXD committed:
Revert protobuf upgrade

U8NWXD committed:
Fix build.py backend test

U8NWXD committed:
Fix Coverage by Removing Unnecessary if

Protobuf is installed in all environments now as part of our development
tools, so there's no need to check whether a google package is in
Python's module path.

U8NWXD committed:
Increase core/tests/test_utils.py coverage

U8NWXD committed:
Merge branch 'develop' into upgrade-pypi-dev

U8NWXD committed:
Remove unused call_through arg to swap_with_call_counter

U8NWXD committed:
Add backend tests for core/test/test_utils.py

U8NWXD committed:
Remove unused GenericTestBase creation funcs

U8NWXD committed:
Remove unused imports

U8NWXD committed:
Add types to mock function

U8NWXD committed:
Remove unused test_utils.py code

U8NWXD committed:
Merge branch 'develop' into upgrade-pypi-dev

U8NWXD committed:
Add test to use unused test_utils.py functionality

U8NWXD committed:
Fix coverage errors

U8NWXD commented:
@seanlip @aasiffaizal @lkbhitesh07 could you please take another look at this PR? I've had to make a lot of changes since you last reviewed

lkbhitesh07 reviewed (approved):
Thanks for the changes @U8NWXD, LGTM for the codeowners files.

seanlip commented on .pylintrc:
```diff
@@ -45,13 +45,17 @@ module-rgx=[A-Za-z_][A-Za-z0-9_]*$
 no-docstring-rgx=test_[a-z_]*|[A-Za-z]*Tests|Mock[A-Za-z]*|mock_[a-z_]*|setUp|tearDown|__.*__
 
+# We allow more flexible type names than pylint normally allows. We allow
+# CamelCased types, with an optional underscore prefix.
```
Maybe also explain what Pylint normally allows?

U8NWXD replied:
done

seanlip commented on core/controllers/profile.py:
```diff
@@ -69,8 +69,8 @@ def get(self, username: str) -> None:
             user_settings.user_id)
-        is_already_subscribed = (self.user_id in subscriber_ids)
-        is_user_visiting_own_profile = (self.user_id == user_settings.user_id)
+        is_already_subscribed = self.user_id in subscriber_ids
```
Out of interest, are the changes in this file forced by the linter?

I actually think the old presentation is clearer (and less error prone, especially in the case of A = B == C) but my opinion here isn't super-strong.

U8NWXD replied:
Yes, the linter now enforces the removal of unnecessary parentheses

seanlip commented on core/domain/user_services.py:
```diff
@@ -74,6 +74,9 @@
 USERNAME_FOR_USER_BEING_DELETED: Final = 'UserBeingDeleted'
 
+# Timeout in seconds for requests.
+TIMEOUT = 60
```
TIMEOUT_SECS

U8NWXD replied:
done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -1493,26 +1488,20 @@ def swap_with_call_counter(
             raises: Exception|None. The exception raised by the swapped
                 function. If None, then no exception is raised.
             returns: *. The return value of the swapped function.
-            call_through: bool. Whether to call through to the real function,
```
I don't mind this getting dropped, but was just curious about why?

U8NWXD replied:
It wasn't being used, so the coverage checks flagged it as not being covered by the tests. Rather than write tests for it or write tests that use it, I dropped it since it wasn't used

seanlip commented on core/tests/test_utils.py:
```diff
@@ -3232,12 +3208,17 @@ def traverse_schema_and_assign_content_ids(
                 for x in value:
                     traverse_schema_and_assign_content_ids(
                         x, schema['items'], ca_name)
+            # We exclude this elif block from coverage because no interaction
```
I think it's fine to actually just do an `else: raise Exception` here, because we won't introduce an unstructured dict type in the future now that we're typing everything fully.

U8NWXD replied:
done

seanlip commented on core/tests/test_utils.py:
```diff
@@ -3609,151 +3590,6 @@ def save_new_story(
         story_services.save_new_story(owner_id, story)
         return story
 
-    def save_new_story_with_story_contents_schema_v1(
```
Were we just not using all these functions?

U8NWXD replied:
correct

seanlip commented on core/utils_test.py:
```diff
@@ -892,7 +892,7 @@ def test_generate_new_session_id(self) -> None:
         with self.assertRaisesRegex(
             utils.ValidationError,
-            'The length of the exploration title should be between 1 and 50 ' 'characters; received '):   # pylint: disable=line-too-long
+            'The length of the exploration title should be between 1 and 50 characters; received '):   # pylint: disable=line-too-long
```
I think you don't need the line-too-long pragma if you just split into two lines (i.e. keep the quotes that were deleted and put a newline after the first deleted quote.

U8NWXD replied:
done

seanlip commented on scripts/build_test.py:
```diff
@@ -1262,7 +1262,15 @@ def test_build_js_files_in_dev_mode_with_exception_raised(self) -> None:
         self.exit_stack.enter_context(self.swap_with_checks(
-            sys, 'exit', lambda _: None, expected_args=[(return_code,)]))
+            sys, 'exit', lambda _: None,
+            expected_args=[
```
Why did this change? Maybe a comment explaining the expected value might be useful.

U8NWXD replied:
I removed an unreachable `return` statement from `run_webpack_compilation` in `scripts/build.py` ([diff](https://github.com/oppia/oppia/pull/18165/files#diff-c786cc346fc8fdc17f1c85f11bc187c153a3511ae549390185c7c39343d2c95dL214)). The return statement was unreachable because of a `sys.exit` call, but in the test, `sys.exit` is mocked, so now it gets called multiple times. Will add a comment explaining this

seanlip commented on scripts/linters/python_linter_test.py:
```diff
@@ -69,8 +70,12 @@ def test_valid_job_imports(self) -> None:
+                batch_jobs_dir, VALID_PY_JOBS_FILEPATH)
         )
+        print('DEBUGGING')
+        print(lint_task_report.get_report())
```
Should this be here? (If so, should the print statements above/below this line be more descriptive?)

U8NWXD replied:
nope--removed

seanlip commented on scripts/populate_sample_contributor_data.py:
```diff
@@ -69,6 +69,9 @@
 CLASSROOM_URL_FRAGMENT: Final = 'math'
 
+# Timeout for requests in seconds.
+TIMEOUT = 60
```
TIMEOUT_SECS

U8NWXD replied:
done

seanlip reviewed (changes requested):
Thanks @U8NWXD! Took a pass -- just minor comments, but generally LGTM. PTAL.

U8NWXD committed:
Address review comments

U8NWXD committed:
Merge branch 'develop' into upgrade-pypi-dev

U8NWXD commented:
@seanlip PTAL

seanlip commented on .pylintrc:
```diff
@@ -46,7 +46,8 @@ module-rgx=[A-Za-z_][A-Za-z0-9_]*$
 # We allow more flexible type names than pylint normally allows. We allow
-# CamelCased types, with an optional underscore prefix.
+# CamelCased types, with an optional underscore prefix. Pylint pnormally
+# rohibits type names from ending in 'Type'.
```
Typo: "normally prohibits"

U8NWXD replied:
done

seanlip reviewed (approved):
Thanks, LGTM! Just one typo.

U8NWXD committed:
Fix typo

U8NWXD committed:
Fix backend test

U8NWXD committed:
Merge branch 'develop' into upgrade-pypi-dev
